Lib/concurrent/futures/_base.py

```diff
@@ -544,18 +545,40 @@ def map(self, fn, *iterables, timeout=None, chunksize=1):
         """
         if timeout is not None:
             end_time = timeout + time.time()
+        if prefetch is None:
+            prefetch = self._max_workers
+        if prefetch < 0:
+            raise ValueError("prefetch count may not be negative")
 
-        fs = [self.submit(fn, *args) for args in zip(*iterables)]
+        argsiter = zip(*iterables)
+
+        fs = collections.deque(self.submit(fn, *args) for args in itertools.islice(argsiter, self._max_workers + prefetch))
 
         # Yield must be hidden in closure so that the futures are submitted
         # before the first iterator value is required.
         def result_iterator():
+            nonlocal argsiter
             try:
-                for future in fs:
+                while fs:
                     if timeout is None:
-                        yield future.result()
+                        res = fs[0].result()
                     else:
-                        yield future.result(end_time - time.time())
+                        res = fs[0].result(end_time - time.time())
+
+                    # Got a result, future needn't be cancelled
+                    del fs[0]
+
```

Comment:
`fs.popleft()` is nicer to read (it's probably identical in performance to `del fs[0]`, but there's a chance it may even be more efficient).

Reply:
Ah, I avoided using `popleft` at the bottom because I view it as specifically for when you want both behaviors (both removing *and* getting the leftmost item); since we only wanted to remove, not get the item, I used the tool that performs only that task.

There is actually a small performance penalty for using `fs.popleft()` over `del fs[0]` (this is the case for most comparisons between otherwise equivalent techniques using method or function calls or a Python syntax feature; the syntax feature's use of dedicated byte codes rather than complicated and relatively expensive generalized function call code paths), but I'll admit it's pretty meaningless in the context of anything as complicated as an Executor (`popleft` takes roughly twice as long, but this is a matter of ~30 ns for `del` vs. ~60 ns for `popleft`).